timer/timer.go

```diff
@@ -0,0 +1,120 @@
+package timer
+
+import (
+	"errors"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+	"github.com/nivista/steady/.gen/protos/common"
+	"github.com/nivista/steady/internal/.gen/protos/messaging"
+)
+
+type (
+	// Timer represents the configuration for the execution of a recurring task.
+	Timer struct {
+		executer  executer
+		scheduler scheduler
+	}
+
+	// Progress records the progress of a recurring task.
+	Progress struct {
+		// CompletedExecutions indicated the number of successfully completed executions.
+		CompletedExecutions int
+
+		// LastExecution indicates which execution was recorded last.
+		LastExecution time.Time
+	}
+
+	executer interface {
+		execute()
+	}
+
+	scheduler interface {
```

Comment:
I think you will want both scheduler and executor to have a validate method on them. And you want timer to have a validate method on it. The timer's validate method will invoke the scheduler and executor validate method.

Reply:
Ok. Validation has to happen on the webservice end so this would change a lot. Basically I'd have to go from services.CreateTimerRequest proto to in memory timer, call validate, and then from there to messaging.CreateTimer proto. Could I do validation directly on the proto objects? If I did, where would that go?